src/appendable_data.rs

```diff
@@ -0,0 +1,542 @@
+    /// The current index of the data when this permission change happened
+    data_index: u64,
+    /// The current index of the owners when this permission change happened
+    owner_entry_index: u64,
+}
+
+impl Permissions for UnpubPermissions {
+    fn is_action_allowed(&self, requester: PublicKey, action: Action) -> bool {
+        match self.permissions.get(&requester) {
+            Some(perms) => perms.is_allowed(action),
+            None => false,
+        }
+    }
+
+    fn data_index(&self) -> u64 {
+        self.data_index
+    }
+
+    fn owner_entry_index(&self) -> u64 {
+        self.owner_entry_index
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct PubPermissions {
+    permissions: BTreeMap<User, PubPermissionSet>,
+    /// The current index of the data when this permission change happened
+    data_index: u64,
+    /// The current index of the owners when this permission change happened
+    owner_entry_index: u64,
+}
+
+impl PubPermissions {
+    fn check_anyone_permissions(&self, action: Action) -> bool {
+        match self.permissions.get(&User::Anyone) {
+            None => false,
+            Some(perms) => perms.is_allowed(action).unwrap_or(false),
+        }
+    }
+}
+
+impl Permissions for PubPermissions {
+    fn is_action_allowed(&self, requester: PublicKey, action: Action) -> bool {
+        match self.permissions.get(&User::Key(requester)) {
+            Some(perms) => perms
+                .is_allowed(action)
+                .unwrap_or_else(|| self.check_anyone_permissions(action)),
+            None => self.check_anyone_permissions(action),
+        }
+    }
+
+    fn data_index(&self) -> u64 {
+        self.data_index
+    }
+
+    fn owner_entry_index(&self) -> u64 {
+        self.owner_entry_index
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct Owners {
+    owners: PublicKeySet,
+    /// The current index of the data when this ownership change happened
+    data_index: u64,
+    /// The current index of the permissions when this ownership change happened
+    permission_entry_index: u64,
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
+struct AppendOnly<P: Permissions> {
+    name: XorName,
+    tag: u64,
+    data: Vec<(Vec<u8>, Vec<u8>)>,
+    permissions: Vec<P>,
+    owners: Vec<Owners>,
+}
+
+/// Common methods for all `AppendOnlyData` flavours.
+pub trait AppendOnlyData<P> {
+    // /// Get a list of permissions for the provided user from the last entry in the permissions list.
+    // fn user_permissions(&self, user: &User) -> Result<&PubPermissionSet, ClientError>;
+
+    /// Return a value for the given key (if it is present).
+    fn get(&self, key: &[u8]) -> Option<&Vec<u8>>;
+
+    /// Get a list of keys and values with the given indexes.
+    fn at(&self, start: Index, end: Index) -> Option<&[(Vec<u8>, Vec<u8>)]>;
```

Comment:
Should we consider changing the name of this function?
`at` seems like getting an entry _at_ a particular index. Maybe we can consider something like `in_range` or `between` or `from_range`

Reply:
Agreed, and I also just spotted "indexes" which should be spelled "indices" :smile: Thanks!